File: gopher-mcp/sdk/go/src/types/filter_types.go
// Package types provides core type definitions for the MCP Filter SDK.
package types

import (
	"fmt"
	"sync"
	"time"
)

// FilterStatus represents the result status of a filter's processing operation.
// It determines how the filter chain should proceed after processing.
type FilterStatus int

const (
	// Continue indicates the filter processed successfully and the chain should continue.
	// The next filter in the chain will receive the processed data.
	Continue FilterStatus = iota

	// StopIteration indicates the filter processed successfully but the chain should stop.
	// No further filters will be executed, and the current result will be returned.
	StopIteration

	// Error indicates the filter encountered an error during processing.
	// The chain will stop and return the error unless configured to bypass errors.
	Error

	// NeedMoreData indicates the filter needs more data to complete processing.
	// Used for filters that work with streaming or chunked data.
	NeedMoreData

	// Buffered indicates the filter has buffered the data for later processing.
	// The chain may continue with empty data or wait based on configuration.
	Buffered
)

// String returns a human-readable string representation of the FilterStatus.
func (s FilterStatus) String() string {
	switch s {
	case Continue:
		return "Continue"
	case StopIteration:
		return "StopIteration"
	case Error:
		return "Error"
	case NeedMoreData:
		return "NeedMoreData"
	case Buffered:
		return "Buffered"
	default:
		return fmt.Sprintf("FilterStatus(%d)", s)
	}
}

// IsTerminal returns true if the status indicates chain termination.
func (s FilterStatus) IsTerminal() bool {
	return s == StopIteration || s == Error
}

// IsSuccess returns true if the status indicates successful processing.
func (s FilterStatus) IsSuccess() bool {
	return s == Continue || s == StopIteration || s == Buffered
}

// FilterPosition indicates where a filter should be placed in a chain.
// It determines the relative position when adding filters dynamically.
type FilterPosition int

const (
	// First indicates the filter should be placed at the beginning of the chain.
	First FilterPosition = iota

	// Last indicates the filter should be placed at the end of the chain.
	Last

	// Before indicates the filter should be placed before a specific filter.
	// Requires a reference filter name or ID.
	Before

	// After indicates the filter should be placed after a specific filter.
	// Requires a reference filter name or ID.
	After
)

// String returns a human-readable string representation of the FilterPosition.
func (p FilterPosition) String() string {
	switch p {
	case First:
		return "First"
	case Last:
		return "Last"
	case Before:
		return "Before"
	case After:
		return "After"
	default:
		return fmt.Sprintf("FilterPosition(%d)", p)
	}
}

// IsValid validates that the position is within the valid range.
func (p FilterPosition) IsValid() bool {
	return p >= First && p <= After
}

// RequiresReference returns true if the position requires a reference filter.
func (p FilterPosition) RequiresReference() bool {
	return p == Before || p == After
}

// FilterError represents specific error codes for filter operations.
// These codes provide detailed information about filter failures.
type FilterError int

const (
	// InvalidConfiguration indicates the filter configuration is invalid.
	InvalidConfiguration FilterError = 1001

	// FilterNotFound indicates the specified filter was not found in the chain.
	FilterNotFound FilterError = 1002

	// FilterAlreadyExists indicates a filter with the same name already exists.
	FilterAlreadyExists FilterError = 1003

	// InitializationFailed indicates the filter failed to initialize.
	InitializationFailed FilterError = 1004

	// ProcessingFailed indicates the filter failed during data processing.
	ProcessingFailed FilterError = 1005

	// ChainProcessingError indicates an error in the filter chain execution.
	ChainProcessingError FilterError = 1006

	// BufferOverflow indicates the buffer size limit was exceeded.
	BufferOverflow FilterError = 1007

	// Timeout indicates the operation exceeded the time limit.
	Timeout FilterError = 1010

	// ResourceExhausted indicates system resources were exhausted.
	ResourceExhausted FilterError = 1011

	// TooManyRequests indicates rate limiting was triggered.
	TooManyRequests FilterError = 1018

	// AuthenticationFailed indicates authentication failed.
	AuthenticationFailed FilterError = 1019

	// ServiceUnavailable indicates the service is temporarily unavailable.
	ServiceUnavailable FilterError = 1021
)

// Error implements the error interface for FilterError.
func (e FilterError) Error() string {
	switch e {
	case InvalidConfiguration:
		return "invalid filter configuration"
	case FilterNotFound:
		return "filter not found"
	case FilterAlreadyExists:
		return "filter already exists"
	case InitializationFailed:
		return "filter initialization failed"
	case ProcessingFailed:
		return "filter processing failed"
	case ChainProcessingError:
		return "filter chain error"
	case BufferOverflow:
		return "buffer overflow"
	case Timeout:
		return "operation timeout"
	case ResourceExhausted:
		return "resource exhausted"
	case TooManyRequests:
		return "too many requests"
	case AuthenticationFailed:
		return "authentication failed"
	case ServiceUnavailable:
		return "service unavailable"
	default:
		return fmt.Sprintf("filter error: %d", e)
	}
}

// String returns a human-readable string representation of the FilterError.
func (e FilterError) String() string {
	return e.Error()
}

// Code returns the numeric error code.
func (e FilterError) Code() int {
	return int(e)
}

// IsRetryable returns true if the error is potentially retryable.
func (e FilterError) IsRetryable() bool {
	switch e {
	case Timeout, ResourceExhausted, TooManyRequests, ServiceUnavailable:
		return true
	default:
		return false
	}
}

// FilterLayer represents the OSI layer at which a filter operates.
// This helps organize filters by their processing level.
type FilterLayer int

const (
	// Transport represents OSI Layer 4 (Transport Layer).
	// Handles TCP, UDP, and other transport protocols.
	Transport FilterLayer = 4

	// Session represents OSI Layer 5 (Session Layer).
	// Manages sessions and connections between applications.
	Session FilterLayer = 5

	// Presentation represents OSI Layer 6 (Presentation Layer).
	// Handles data encoding, encryption, and compression.
	Presentation FilterLayer = 6

	// Application represents OSI Layer 7 (Application Layer).
	// Processes application-specific protocols like HTTP, gRPC.
	Application FilterLayer = 7

	// Custom represents a custom layer outside the OSI model.
	// Used for filters that don't fit standard layer classifications.
	Custom FilterLayer = 99
)

// String returns a human-readable string representation of the FilterLayer.
func (l FilterLayer) String() string {
	switch l {
	case Transport:
		return "Transport (L4)"
	case Session:
		return "Session (L5)"
	case Presentation:
		return "Presentation (L6)"
	case Application:
		return "Application (L7)"
	case Custom:
		return "Custom"
	default:
		return fmt.Sprintf("FilterLayer(%d)", l)
	}
}

// IsValid validates that the layer is a recognized value.
func (l FilterLayer) IsValid() bool {
	return l == Transport || l == Session || l == Presentation || l == Application || l == Custom
}

// OSILayer returns the OSI model layer number (4-7) or 0 for custom.
func (l FilterLayer) OSILayer() int {
	if l >= Transport && l <= Application {
		return int(l)
	}
	return 0
}

// FilterConfig contains configuration settings for a filter.
// It provides all necessary parameters to initialize and operate a filter.
type FilterConfig struct {
	// Name is the unique identifier for the filter instance.
	Name string `json:"name"`

	// Type specifies the filter type (e.g., "http", "auth", "log").
	Type string `json:"type"`

	// Settings contains filter-specific configuration as key-value pairs.
	Settings map[string]interface{} `json:"settings,omitempty"`

	// Layer indicates the OSI layer at which the filter operates.
	Layer FilterLayer `json:"layer"`

	// Enabled determines if the filter is active in the chain.
	Enabled bool `json:"enabled"`

	// Priority determines the filter's execution order (lower = higher priority).
	Priority int `json:"priority"`

	// TimeoutMs specifies the maximum processing time in milliseconds.
	TimeoutMs int `json:"timeout_ms"`

	// BypassOnError allows the chain to continue if this filter fails.
	BypassOnError bool `json:"bypass_on_error"`

	// MaxBufferSize sets the maximum buffer size in bytes.
	MaxBufferSize int `json:"max_buffer_size"`

	// EnableStatistics enables performance metrics collection.
	EnableStatistics bool `json:"enable_statistics"`
}

// Validate checks if the FilterConfig contains valid values.
// It returns a slice of errors for all validation failures found.
func (c *FilterConfig) Validate() []error {
	var errors []error

	// Check Name is not empty
	if c.Name == "" {
		errors = append(errors, fmt.Errorf("filter name cannot be empty"))
	}

	// Check Type is not empty
	if c.Type == "" {
		errors = append(errors, fmt.Errorf("filter type cannot be empty"))
	}

	// Check MaxBufferSize is positive if set
	if c.MaxBufferSize < 0 {
		errors = append(errors, fmt.Errorf("max buffer size cannot be negative: %d", c.MaxBufferSize))
	}
	if c.MaxBufferSize == 0 {
		// Set a default if not specified
		c.MaxBufferSize = 1024 * 1024 // 1MB default
	}

	// Check TimeoutMs is non-negative
	if c.TimeoutMs < 0 {
		errors = append(errors, fmt.Errorf("timeout cannot be negative: %d ms", c.TimeoutMs))
	}

	// Check Priority is within reasonable range (0-1000)
	if c.Priority < 0 || c.Priority > 1000 {
		errors = append(errors, fmt.Errorf("priority must be between 0 and 1000, got: %d", c.Priority))
	}

	// Validate Layer if specified
	if c.Layer != 0 && !c.Layer.IsValid() {
		errors = append(errors, fmt.Errorf("invalid filter layer: %d", c.Layer))
	}

	return errors
}

// FilterStatistics tracks performance metrics for a filter.
// All fields should be accessed atomically in concurrent environments.
type FilterStatistics struct {
	// BytesProcessed is the total number of bytes processed by the filter.
	BytesProcessed uint64 `json:"bytes_processed"`

	// PacketsProcessed is the total number of packets/messages processed.
	PacketsProcessed uint64 `json:"packets_processed"`

	// ProcessCount is the total number of times the filter has been invoked.
	ProcessCount uint64 `json:"process_count"`

	// ErrorCount is the total number of errors encountered.
	ErrorCount uint64 `json:"error_count"`

	// ProcessingTimeUs is the total processing time in microseconds.
	ProcessingTimeUs uint64 `json:"processing_time_us"`

	// AverageProcessingTimeUs is the average processing time per invocation.
	AverageProcessingTimeUs float64 `json:"average_processing_time_us"`

	// MaxProcessingTimeUs is the maximum processing time recorded.
	MaxProcessingTimeUs uint64 `json:"max_processing_time_us"`

	// MinProcessingTimeUs is the minimum processing time recorded.
	MinProcessingTimeUs uint64 `json:"min_processing_time_us"`

	// CurrentBufferUsage is the current buffer memory usage in bytes.
	CurrentBufferUsage uint64 `json:"current_buffer_usage"`

	// PeakBufferUsage is the peak buffer memory usage in bytes.
	PeakBufferUsage uint64 `json:"peak_buffer_usage"`

	// ThroughputBps is the current throughput in bytes per second.
	ThroughputBps float64 `json:"throughput_bps"`

	// ErrorRate is the percentage of errors (0-100).
	ErrorRate float64 `json:"error_rate"`

	// CustomMetrics allows filters to store custom metrics.
	CustomMetrics map[string]interface{} `json:"custom_metrics,omitempty"`
}

// String returns a human-readable summary of the filter statistics.
func (s *FilterStatistics) String() string {
	return fmt.Sprintf(
		"FilterStats{Processed: %d bytes/%d packets, Invocations: %d, Errors: %d, "+
			"AvgTime: %.2fμs, MaxTime: %dμs, MinTime: %dμs, "+
			"BufferUsage: %d/%d bytes, Throughput: %.2f B/s}",
		s.BytesProcessed, s.PacketsProcessed, s.ProcessCount, s.ErrorCount,
		s.AverageProcessingTimeUs, s.MaxProcessingTimeUs, s.MinProcessingTimeUs,
		s.CurrentBufferUsage, s.PeakBufferUsage, s.ThroughputBps,
	)
}

// FilterResult represents the result of a filter's processing operation.
// It contains the processing status, output data, and metadata.
type FilterResult struct {
	// Status indicates the result of the filter processing.
	Status FilterStatus `json:"status"`

	// Data contains the processed output data.
	Data []byte `json:"data,omitempty"`

	// Error contains any error that occurred during processing.
	Error error `json:"error,omitempty"`

	// Metadata contains additional information about the processing.
	Metadata map[string]interface{} `json:"metadata,omitempty"`

	// StartTime marks when processing began.
	StartTime time.Time `json:"start_time"`

	// EndTime marks when processing completed.
	EndTime time.Time `json:"end_time"`

	// StopChain indicates if the filter chain should stop after this filter.
	StopChain bool `json:"stop_chain"`

	// SkipCount indicates how many filters to skip in the chain.
	SkipCount int `json:"skip_count"`
}

// Duration calculates the processing time for this result.
func (r *FilterResult) Duration() time.Duration {
	if r.EndTime.IsZero() || r.StartTime.IsZero() {
		return 0
	}
	return r.EndTime.Sub(r.StartTime)
}

// IsSuccess returns true if the result indicates successful processing.
// Success is defined as Continue or StopIteration status without errors.
func (r *FilterResult) IsSuccess() bool {
	if r == nil {
		return false
	}
	return (r.Status == Continue || r.Status == StopIteration) && r.Error == nil
}

// IsError returns true if the result indicates an error occurred.
// An error is indicated by Error status or non-nil Error field.
func (r *FilterResult) IsError() bool {
	if r == nil {
		return false
	}
	return r.Status == Error || r.Error != nil
}

// Validate checks the consistency of the FilterResult.
// It ensures status is valid and error fields are consistent.
func (r *FilterResult) Validate() error {
	if r == nil {
		return fmt.Errorf("filter result is nil")
	}

	// Check status is valid
	if r.Status < Continue || r.Status > Buffered {
		return fmt.Errorf("invalid filter status: %d", r.Status)
	}

	// Check error consistency
	if r.Status == Error && r.Error == nil {
		return fmt.Errorf("error status without error field")
	}

	if r.Status != Error && r.Error != nil {
		return fmt.Errorf("non-error status with error field: status=%v, error=%v", r.Status, r.Error)
	}

	// Check data length consistency if metadata present
	if r.Metadata != nil {
		if dataLen, ok := r.Metadata["data_length"].(int); ok {
			if dataLen != len(r.Data) {
				return fmt.Errorf("data length mismatch: metadata=%d, actual=%d", dataLen, len(r.Data))
			}
		}
	}

	return nil
}

// filterResultPool is a pool for reusing FilterResult instances.
var filterResultPool = sync.Pool{
	New: func() interface{} {
		return &FilterResult{
			Metadata: make(map[string]interface{}),
		}
	},
}

// GetResult retrieves a FilterResult from the pool.
// The result is cleared and ready for use.
func GetResult() *FilterResult {
	r := filterResultPool.Get().(*FilterResult)
	r.reset()
	return r
}

// Release returns the FilterResult to the pool.
// All fields are cleared to prevent data leaks.
func (r *FilterResult) Release() {
	if r == nil {
		return
	}
	r.reset()
	filterResultPool.Put(r)
}

// reset clears all fields in the FilterResult.
func (r *FilterResult) reset() {
	r.Status = Continue
	r.Data = nil
	r.Error = nil

	// Clear metadata map
	if r.Metadata == nil {
		r.Metadata = make(map[string]interface{})
	} else {
		for k := range r.Metadata {
			delete(r.Metadata, k)
		}
	}

	r.StartTime = time.Time{}
	r.EndTime = time.Time{}
	r.StopChain = false
	r.SkipCount = 0
}

// Success creates a successful FilterResult with the provided data.
func Success(data []byte) *FilterResult {
	now := time.Now()
	return &FilterResult{
		Status:    Continue,
		Data:      data,
		StartTime: now,
		EndTime:   now,
		Metadata:  make(map[string]interface{}),
	}
}

// Error creates an error FilterResult with the provided error and code.
func ErrorResult(err error, code FilterError) *FilterResult {
	now := time.Now()
	return &FilterResult{
		Status:    Error,
		Error:     fmt.Errorf("%s: %w", code.Error(), err),
		StartTime: now,
		EndTime:   now,
		Metadata: map[string]interface{}{
			"error_code": code.Code(),
		},
	}
}

// ContinueWith creates a FilterResult that continues with the provided data.
func ContinueWith(data []byte) *FilterResult {
	now := time.Now()
	return &FilterResult{
		Status:    Continue,
		Data:      data,
		StartTime: now,
		EndTime:   now,
		Metadata:  make(map[string]interface{}),
	}
}

// Blocked creates a FilterResult indicating the request was blocked.
func Blocked(reason string) *FilterResult {
	now := time.Now()
	return &FilterResult{
		Status:    StopIteration,
		StopChain: true,
		StartTime: now,
		EndTime:   now,
		Metadata: map[string]interface{}{
			"blocked_reason": reason,
		},
	}
}

// StopIterationResult creates a FilterResult that stops the filter chain.
func StopIterationResult() *FilterResult {
	now := time.Now()
	return &FilterResult{
		Status:    StopIteration,
		StopChain: true,
		StartTime: now,
		EndTime:   now,
		Metadata:  make(map[string]interface{}),
	}
}

// FilterEventArgs provides base event arguments for filter events.
// This struct can be embedded in specific event types.
type FilterEventArgs struct {
	// FilterName is the name of the filter that generated the event.
	FilterName string `json:"filter_name"`

	// FilterType is the type of the filter that generated the event.
	FilterType string `json:"filter_type"`

	// Timestamp is when the event occurred.
	Timestamp time.Time `json:"timestamp"`

	// Data contains event-specific data as key-value pairs.
	Data map[string]interface{} `json:"data,omitempty"`
}

// FilterDataEventArgs provides event arguments for filter data processing events.
// It embeds FilterEventArgs and adds data-specific fields.
type FilterDataEventArgs struct {
	// Embed the base event arguments
	FilterEventArgs

	// Buffer contains the data being processed.
	Buffer []byte `json:"buffer,omitempty"`

	// Offset is the starting position in the buffer.
	Offset int `json:"offset"`

	// Length is the number of bytes to process from the offset.
	Length int `json:"length"`

	// Status is the processing status for this data.
	Status FilterStatus `json:"status"`

	// Handled indicates if the event has been handled.
	Handled bool `json:"handled"`
}

// GetData returns the relevant slice of the buffer based on offset and length.
// It handles bounds checking to prevent panics.
func (e *FilterDataEventArgs) GetData() []byte {
	if e.Buffer == nil || e.Offset < 0 || e.Length <= 0 {
		return nil
	}

	// Ensure we don't exceed buffer bounds
	end := e.Offset + e.Length
	if e.Offset >= len(e.Buffer) {
		return nil
	}
	if end > len(e.Buffer) {
		end = len(e.Buffer)
	}

	return e.Buffer[e.Offset:end]
}

File: gopher-mcp/sdk/csharp/src/Types/FilterTypes.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace GopherMcp.Types
{
    /// <summary>
    /// Filter processing status indicating how to continue processing
    /// </summary>
    public enum FilterStatus
    {
        /// <summary>
        /// Continue processing with next filter
        /// </summary>
        Continue = 0,

        /// <summary>
        /// Stop iteration and return
        /// </summary>
        StopIteration = 1,

        /// <summary>
        /// Error occurred during processing
        /// </summary>
        Error = 2,

        /// <summary>
        /// Filter needs more data to proceed
        /// </summary>
        NeedMoreData = 3,

        /// <summary>
        /// Filter has buffered data for later processing
        /// </summary>
        Buffered = 4
    }

    /// <summary>
    /// Filter position in chain
    /// </summary>
    public enum FilterPosition
    {
        /// <summary>
        /// Add filter at the beginning of the chain
        /// </summary>
        First = 0,

        /// <summary>
        /// Add filter at the end of the chain
        /// </summary>
        Last = 1,

        /// <summary>
        /// Add filter before a specific filter
        /// </summary>
        Before = 2,

        /// <summary>
        /// Add filter after a specific filter
        /// </summary>
        After = 3
    }

    /// <summary>
    /// Filter-specific error codes
    /// </summary>
    public enum FilterError
    {
        /// <summary>
        /// No error
        /// </summary>
        None = 0,

        /// <summary>
        /// Invalid filter configuration
        /// </summary>
        InvalidConfiguration = 1001,

        /// <summary>
        /// Filter not found
        /// </summary>
        FilterNotFound = 1002,

        /// <summary>
        /// Filter already exists
        /// </summary>
        FilterAlreadyExists = 1003,

        /// <summary>
        /// Filter initialization failed
        /// </summary>
        InitializationFailed = 1004,

        /// <summary>
        /// Filter processing failed
        /// </summary>
        ProcessingFailed = 1005,

        /// <summary>
        /// Filter chain error
        /// </summary>
        ChainError = 1006,

        /// <summary>
        /// Buffer overflow
        /// </summary>
        BufferOverflow = 1007,

        /// <summary>
        /// Buffer underflow
        /// </summary>
        BufferUnderflow = 1008,

        /// <summary>
        /// Invalid buffer state
        /// </summary>
        InvalidBufferState = 1009,

        /// <summary>
        /// Filter timeout
        /// </summary>
        Timeout = 1010,

        /// <summary>
        /// Resource exhausted
        /// </summary>
        ResourceExhausted = 1011,

        /// <summary>
        /// Not supported
        /// </summary>
        NotSupported = 1012,

        /// <summary>
        /// Permission denied
        /// </summary>
        PermissionDenied = 1013,

        /// <summary>
        /// Filter is in invalid state for operation
        /// </summary>
        InvalidState = 1014,

        /// <summary>
        /// Dependency not met
        /// </summary>
        DependencyNotMet = 1015,

        /// <summary>
        /// Bypass filter
        /// </summary>
        Bypass = 1016,

        /// <summary>
        /// Internal error
        /// </summary>
        InternalError = 1017,

        /// <summary>
        /// Too many requests (rate limit)
        /// </summary>
        TooManyRequests = 1018,

        /// <summary>
        /// Authentication failed
        /// </summary>
        AuthenticationFailed = 1019,

        /// <summary>
        /// Authorization failed
        /// </summary>
        AuthorizationFailed = 1020,

        /// <summary>
        /// Service unavailable
        /// </summary>
        ServiceUnavailable = 1021,

        /// <summary>
        /// Unauthorized access
        /// </summary>
        Unauthorized = 1022,

        /// <summary>
        /// Network error
        /// </summary>
        NetworkError = 1023,

        /// <summary>
        /// Retry exhausted
        /// </summary>
        RetryExhausted = 1024,

        /// <summary>
        /// Not found
        /// </summary>
        NotFound = 404,

        /// <summary>
        /// Forbidden
        /// </summary>
        Forbidden = 403
    }

    /// <summary>
    /// Filter layer in the processing stack
    /// </summary>
    public enum FilterLayer
    {
        /// <summary>
        /// Transport layer (L4)
        /// </summary>
        Transport = 0,

        /// <summary>
        /// Session layer (L5)
        /// </summary>
        Session = 1,

        /// <summary>
        /// Presentation layer (L6)
        /// </summary>
        Presentation = 2,

        /// <summary>
        /// Application layer (L7)
        /// </summary>
        Application = 3,

        /// <summary>
        /// Custom layer
        /// </summary>
        Custom = 99
    }

    /// <summary>
    /// Filter configuration
    /// </summary>
    public class FilterConfig
    {
        /// <summary>
        /// Gets or sets the filter name
        /// </summary>
        public string Name { get; set; }

        /// <summary>
        /// Gets or sets the filter type
        /// </summary>
        public string Type { get; set; }

        /// <summary>
        /// Gets or sets the filter settings as key-value pairs
        /// </summary>
        public Dictionary<string, object> Settings { get; set; }

        /// <summary>
        /// Gets or sets the filter layer
        /// </summary>
        public FilterLayer Layer { get; set; }

        /// <summary>
        /// Gets or sets the memory pool name for this filter
        /// </summary>
        public string MemoryPool { get; set; }

        /// <summary>
        /// Gets or sets whether the filter is enabled
        /// </summary>
        public bool Enabled { get; set; } = true;

        /// <summary>
        /// Gets or sets the filter priority (lower values = higher priority)
        /// </summary>
        public int Priority { get; set; } = 100;

        /// <summary>
        /// Gets or sets the filter timeout in milliseconds
        /// </summary>
        public int TimeoutMs { get; set; } = 30000;

        /// <summary>
        /// Gets or sets whether to bypass this filter on error
        /// </summary>
        public bool BypassOnError { get; set; } = false;

        /// <summary>
        /// Gets or sets the maximum buffer size for this filter
        /// </summary>
        public int MaxBufferSize { get; set; } = 65536;

        /// <summary>
        /// Gets or sets whether to enable statistics for this filter
        /// </summary>
        public bool EnableStatistics { get; set; } = true;

        /// <summary>
        /// Gets or sets the timeout for this filter
        /// </summary>
        public TimeSpan Timeout { get; set; } = TimeSpan.FromSeconds(30);

        /// <summary>
        /// Initializes a new instance of FilterConfig
        /// </summary>
        public FilterConfig()
        {
            Settings = new Dictionary<string, object>();
        }

        /// <summary>
        /// Initializes a new instance of FilterConfig with a name and type
        /// </summary>
        public FilterConfig(string name, string type) : this()
        {
            Name = name;
            Type = type;
        }

        /// <summary>
        /// Clone the configuration
        /// </summary>
        public FilterConfig Clone()
        {
            return new FilterConfig
            {
                Name = Name,
                Type = Type,
                Settings = new Dictionary<string, object>(Settings),
                Layer = Layer,
                MemoryPool = MemoryPool,
                Enabled = Enabled,
                Priority = Priority,
                TimeoutMs = TimeoutMs,
                BypassOnError = BypassOnError,
                MaxBufferSize = MaxBufferSize
            };
        }

        /// <summary>
        /// Sets a configuration setting
        /// </summary>
        /// <param name="key">Setting key</param>
        /// <param name="value">Setting value</param>
        public void SetSetting(string key, object value)
        {
            if (Settings == null)
            {
                Settings = new Dictionary<string, object>();
            }
            Settings[key] = value;
        }

        /// <summary>
        /// Gets a configuration setting
        /// </summary>
        /// <typeparam name="T">Value type</typeparam>
        /// <param name="key">Setting key</param>
        /// <param name="defaultValue">Default value if not found</param>
        /// <returns>Setting value or default</returns>
        public T GetSetting<T>(string key, T defaultValue = default)
        {
            if (Settings != null && Settings.TryGetValue(key, out var value))
            {
                if (value is T typedValue)
                {
                    return typedValue;
                }
            }
            return defaultValue;
        }

        /// <summary>
        /// Validates the configuration
        /// </summary>
        /// <param name="errors">List of validation errors</param>
        /// <returns>True if valid, false otherwise</returns>
        public virtual bool Validate(out List<string> errors)
        {
            errors = new List<string>();

            if (string.IsNullOrWhiteSpace(Name))
            {
                errors.Add("Filter name is required");
            }

            if (string.IsNullOrWhiteSpace(Type))
            {
                errors.Add("Filter type is required");
            }

            if (MaxBufferSize <= 0)
            {
                errors.Add("MaxBufferSize must be greater than 0");
            }

            if (TimeoutMs < 0)
            {
                errors.Add("TimeoutMs cannot be negative");
            }

            return errors.Count == 0;
        }
    }

    /// <summary>
    /// Filter statistics
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public struct FilterStatistics
    {
        /// <summary>
        /// Total bytes processed
        /// </summary>
        public ulong BytesProcessed;

        /// <summary>
        /// Total packets processed
        /// </summary>
        public ulong PacketsProcessed;

        /// <summary>
        /// Total process count
        /// </summary>
        public ulong ProcessCount;

        /// <summary>
        /// Total errors encountered
        /// </summary>
        public ulong ErrorCount;

        /// <summary>
        /// Total processing time in microseconds
        /// </summary>
        public ulong ProcessingTimeUs;

        /// <summary>
        /// Average processing time per packet in microseconds
        /// </summary>
        public double AverageProcessingTimeUs;

        /// <summary>
        /// Maximum processing time in microseconds
        /// </summary>
        public ulong MaxProcessingTimeUs;

        /// <summary>
        /// Minimum processing time in microseconds
        /// </summary>
        public ulong MinProcessingTimeUs;

        /// <summary>
        /// Current buffer usage in bytes
        /// </summary>
        public ulong CurrentBufferUsage;

        /// <summary>
        /// Peak buffer usage in bytes
        /// </summary>
        public ulong PeakBufferUsage;

        /// <summary>
        /// Number of times the filter was bypassed
        /// </summary>
        public ulong BypassCount;

        /// <summary>
        /// Number of timeouts
        /// </summary>
        public ulong TimeoutCount;

        /// <summary>
        /// Throughput in bytes per second
        /// </summary>
        public double ThroughputBps;

        /// <summary>
        /// Get a string representation of the statistics
        /// </summary>
        public override string ToString()
        {
            return $"FilterStatistics: BytesProcessed={BytesProcessed}, PacketsProcessed={PacketsProcessed}, " +
                   $"Errors={ErrorCount}, AvgTime={AverageProcessingTimeUs}μs, Throughput={ThroughputBps:F2}Bps";
        }
    }

    /// <summary>
    /// Base event arguments for filter events
    /// </summary>
    public class FilterEventArgs : EventArgs
    {
        /// <summary>
        /// Gets the filter name
        /// </summary>
        public string FilterName { get; }

        /// <summary>
        /// Gets the filter type
        /// </summary>
        public string FilterType { get; }

        /// <summary>
        /// Gets the timestamp of the event
        /// </summary>
        public DateTime Timestamp { get; }

        /// <summary>
        /// Gets additional event data
        /// </summary>
        public Dictionary<string, object> Data { get; }

        /// <summary>
        /// Initializes a new instance of FilterEventArgs
        /// </summary>
        public FilterEventArgs(string filterName, string filterType)
        {
            FilterName = filterName;
            FilterType = filterType;
            Timestamp = DateTime.UtcNow;
            Data = new Dictionary<string, object>();
        }

        /// <summary>
        /// Initializes a new instance of FilterEventArgs with additional data
        /// </summary>
        public FilterEventArgs(string filterName, string filterType, Dictionary<string, object> data)
            : this(filterName, filterType)
        {
            if (data != null)
            {
                Data = new Dictionary<string, object>(data);
            }
        }
    }

    /// <summary>
    /// Event arguments for filter data events
    /// </summary>
    public class FilterDataEventArgs : FilterEventArgs
    {
        /// <summary>
        /// Gets the data buffer
        /// </summary>
        public byte[] Buffer { get; }

        /// <summary>
        /// Gets the data offset
        /// </summary>
        public int Offset { get; }

        /// <summary>
        /// Gets the data length
        /// </summary>
        public int Length { get; }

        /// <summary>
        /// Gets or sets the processing status
        /// </summary>
        public FilterStatus Status { get; set; }

        /// <summary>
        /// Gets or sets whether the event was handled
        /// </summary>
        public bool Handled { get; set; }

        /// <summary>
        /// Initializes a new instance of FilterDataEventArgs
        /// </summary>
        public FilterDataEventArgs(string filterName, string filterType, byte[] buffer, int offset, int length)
            : base(filterName, filterType)
        {
            Buffer = buffer;
            Offset = offset;
            Length = length;
            Status = FilterStatus.Continue;
            Handled = false;
        }

        /// <summary>
        /// Get the data as a span
        /// </summary>
        public ReadOnlySpan<byte> GetData()
        {
            return new ReadOnlySpan<byte>(Buffer, Offset, Length);
        }
    }

    /// <summary>
    /// Filter processing result
    /// </summary>
    public class FilterResult
    {
        /// <summary>
        /// Gets or sets the processing status
        /// </summary>
        public FilterStatus Status { get; set; }

        /// <summary>
        /// Gets or sets the output data
        /// </summary>
        public byte[] Data { get; set; }

        /// <summary>
        /// Gets or sets the output data offset
        /// </summary>
        public int Offset { get; set; }

        /// <summary>
        /// Gets or sets the output data length
        /// </summary>
        public int Length { get; set; }

        /// <summary>
        /// Gets or sets the error code if Status is Error
        /// </summary>
        public FilterError ErrorCode { get; set; }

        /// <summary>
        /// Gets or sets the error message if Status is Error
        /// </summary>
        public string ErrorMessage { get; set; }

        /// <summary>
        /// Gets or sets additional result metadata
        /// </summary>
        public Dictionary<string, object> Metadata { get; set; }

        /// <summary>
        /// Gets whether the result indicates success
        /// </summary>
        public bool IsSuccess => Status == FilterStatus.Continue || Status == FilterStatus.StopIteration;

        /// <summary>
        /// Gets whether the result indicates an error
        /// </summary>
        public bool IsError => Status == FilterStatus.Error;

        /// <summary>
        /// Initializes a new instance of FilterResult
        /// </summary>
        public FilterResult()
        {
            Status = FilterStatus.Continue;
            Metadata = new Dictionary<string, object>();
        }

        /// <summary>
        /// Initializes a new instance of FilterResult with status
        /// </summary>
        public FilterResult(FilterStatus status) : this()
        {
            Status = status;
        }

        /// <summary>
        /// Initializes a new instance of FilterResult with status and data
        /// </summary>
        public FilterResult(FilterStatus status, byte[] data, int offset, int length) : this(status)
        {
            Data = data;
            Offset = offset;
            Length = length;
        }

        /// <summary>
        /// Create a success result
        /// </summary>
        public static FilterResult Success()
        {
            return new FilterResult(FilterStatus.Continue);
        }

        /// <summary>
        /// Create a success result with data
        /// </summary>
        public static FilterResult Success(byte[] data, int offset, int length)
        {
            return new FilterResult(FilterStatus.Continue, data, offset, length);
        }

        /// <summary>
        /// Create an error result
        /// </summary>
        public static FilterResult Error(FilterError errorCode, string message = null)
        {
            return new FilterResult(FilterStatus.Error)
            {
                ErrorCode = errorCode,
                ErrorMessage = message
            };
        }

        /// <summary>
        /// Create an error result with just a message (uses ProcessingFailed as default error code)
        /// </summary>
        public static FilterResult Error(string message)
        {
            return new FilterResult(FilterStatus.Error)
            {
                ErrorCode = FilterError.ProcessingFailed,
                ErrorMessage = message
            };
        }

        /// <summary>
        /// Create an error result with message and error code (backward compatibility)
        /// </summary>
        public static FilterResult Error(string message, FilterError errorCode)
        {
            return new FilterResult(FilterStatus.Error)
            {
                ErrorCode = errorCode,
                ErrorMessage = message
            };
        }

        /// <summary>
        /// Create a stop iteration result
        /// </summary>
        public static FilterResult StopIteration()
        {
            return new FilterResult(FilterStatus.StopIteration);
        }

        /// <summary>
        /// Create a continue result (pass-through)
        /// </summary>
        public static FilterResult Continue()
        {
            return new FilterResult(FilterStatus.Continue);
        }

        /// <summary>
        /// Create a continue result with data
        /// </summary>
        public static FilterResult Continue(byte[] data)
        {
            return new FilterResult(FilterStatus.Continue, data, 0, data?.Length ?? 0);
        }
    }
}
